include/revng/Pipeline/ContainerInspector.h

```diff
@@ -0,0 +1,115 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/ManagedStatic.h"
+
+#include "revng/Pipeline/Containers.h"
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/Target.h"
+
+namespace Pipeline {
+
+/// If a downstream user of reveng wishes to extend the enumeration and
+/// removal rule from a container it can do so by extending this class,
+/// specialized with the type of the container in question.
+///
+/// the enumerate and removeTarget methods must be implemented with the
+/// appropriate rules
+///
+///
+template<typename Container>
+class ContainerInspector {
+private:
+  Kind *K;
+
+public:
+  using StaticContainer = llvm::SmallVector<ContainerInspector *, 4>;
+
+  ContainerInspector(Kind &K) : K(&K) {
+    getRegisteredInspectors().push_back(this);
+  }
+  virtual ~ContainerInspector() = default;
+
+  virtual TargetsList
+  enumerate(const Context &Ctx, const Container &ToInspect) const = 0;
+  virtual bool remove(const Context &Ctx,
+                      const TargetsList &Targets,
+                      Container &ToInspect) const = 0;
+
+  static TargetsList
+  enumerateTargets(const Context &Ctx, const Container &ToInspect) {
+    TargetsList ToReturn;
+    for (const ContainerInspector *Inspector : getRegisteredInspectors())
+      ToReturn.merge(Inspector->enumerate(Ctx, ToInspect));
+
+    return ToReturn;
+  }
+
+  static bool removeTarget(const Context &Ctx,
+                           const TargetsList &Targets,
+                           Container &ToInspect) {
+    bool RemovedAll = true;
+    for (const ContainerInspector *Inspector : getRegisteredInspectors())
+      RemovedAll = Inspector->remove(Ctx,
+                                     Targets.filter(*Inspector->K),
+                                     ToInspect)
+                   and RemovedAll;
+
+    return RemovedAll;
+  }
+
+private:
+  static StaticContainer &getRegisteredInspectors() {
+    static StaticContainer List;
+    return List;
+  }
+};
+
+/// A inspectable container is a container that delegates the enumeration and
+/// remova of its contents to inspectors, which can be dinamically registered
+/// at runtime.
```

Comment:
This documentation seems to suggest that a non-inspectable container *does not* delegate and does the thing by itself.
Is this the case or a non-inspectable container is simply... non-inspectable at all?
Also:
```suggestion
/// An inspectable container is a container that delegates the enumeration and
/// removal of its contents to inspectors, which can be dynamically registered
/// at runtime.
```

Reply:
yeah, it is inspectable in the sense that there exists inspectors inspecting it

Reply:
I asked about non-inspectable ones.

Which of the following?
1. Non-inspectable containers impossible to inspect (i.e., enumerate their content), or
2. Non-inspectable containers can enumerate their content autonomously, no need for an inspector

Reply:
2, non inspectable containers must provide their own enumerate instead of delegating it to the inspectors

Reply:
OK, please add a line to clarify that non-inspectable containers are "self-aware", they can enumerate themselves.
Maybe we should have called them `ExternallyInspectableContainer`s, but OK.